docs(treelist): Add Cell Selection documentation and revamp Selection… (#2302)
telerik/blazor-docs: components/treelist/selection/cells.md

---
title: Cell Selection
page_title: TreeList - Cells Selection
description: Learn how to select cell in Blazor TreeList component. Explore the selected cells. Discover cell selection bevahior when combined with other TreeList features. Try the practical sample code for cell selection.
slug: treelist-selection-cell
tags: telerik,blazor,treelist,selection,cells
position: 5
---

# Cell Selection

The TreeList component supports [single or multiple cell selection]({%slug treelist-selection-overview%}#use-single-or-multiple-selection). You can select a cell with mouse click anywhere in the cell. You can access the collection of selected cells, use this collection and manipulate it. You can subscribe to selection events.

## Basics

To select a cell, click anywhere in it.

> To select a range of cells in one or more columns, hold the **Shift** key, while clicking on the first and last cell of the range. To select or deselect multiple cells that don't belong to a range, hold the **Ctrl** key.

Reviewer: Generally, our style guide suggests monospace formatting for key names. [Source.](https://docs.telerik.com/style-guide/element-formatting)

The `SelectedCellsChanged` event handler cannot be awaited. To execute asynchronous operations when the user selects rows, use the [`OnRowClick`]({%slug treelist-events%}#onrowclick) or [`OnRowDoubleClick`]({%slug treelist-events%}#onrowdoubleclick) event instead.

## TreeListSelectedCellDescriptor

The `TreeListSelectedCellDescriptor` type exposes the following properties:

@[template](/_contentTemplates/common/parameters-table-styles.md#table-layout)

| Property Name | Type | Description |
| --- | --- | --- |
| `ColumnField` | `string` | The value of the TreeList column `Field` parameter, if set. |
| `ColumnId` | `string` | The value of the TreeList column `Id` parameter, if set. |
| `DataItem` | `object` | The TreeList data item instance. Cast it to the actual TreeList model type before use. |

## Selection When Data Changes

When the TreeList `Data` collection changes, the `SelectedCells` collection has the following behavior:

* When the user updates a selected cell and the item instance is replaced, you have to also replace the `DataItem` object in the `SelectedCells` collection. Do that in the [TreeList `OnUpdate` event]({%slug treelist-editing-overview%}#events).

> * When the user deletes a row with selected cells, update the `SelectedCells` collection in the the TreeList `OnDelete` event handler.

Reviewer: ```suggestion
* When the user deletes a row with selected cells, update the `SelectedCells` collection in the TreeList `OnDelete` event handler.
```

* To select cells from a new item in the TreeList you can use the [`OnCreate` event]({%slug treelist-editing-overview%}#events) to update the `SelectedCells` collection.

### Equals Comparison

The items in `SelectedCells` are compared against the items in the TreeList data in order to determine which cells will be highlighted. The default framework behavior is to compare objects by reference. The data item references may not match when:

> * The `SelectedCells` are obtained from a different data source than the all TreeList items, for example, from a separate service.

Reviewer: Generally speaking, single item lists are not really lists and should be avoided. What's the case for a single item list here? Is it better visibility and rendering on the page?

```suggestion
* The `SelectedCells` are obtained from a different data source than all the TreeList items, for example, from a separate service.
```

In such cases, the selected cells may not appear as expected. You have to [override the `Equals` method of the TreeList model class]({%slug treelist-state%}#basics) so that the items are compared by a unique identifier rather than by reference. When you override `Equals`, it is also recommended to override the [`GetHashCode`](https://docs.microsoft.com/en-us/dotnet/api/system.object.gethashcode) method.

## Cell Selection and Other TreeList Features

> The selection feature behavior may vary when the TreeList configuration combines cell selection and other TreeList features, such as editing, virtualization, paging, templates. In such cases you need to consider certain limitation or include some modications.

Reviewer: ```suggestion
The selection feature behavior may vary when the TreeList configuration combines cell selection and other TreeList features, such as editing, virtualization, paging, templates. In such cases you need to consider certain limitations or include some modifications.
```

* Cell selection is not supported with [`Incell` edit mode]({%slug treelist-editing-incell%}) due to the overlapping pointer events that trigger selection and editing. If both features are enabled, only the editing will work.
* [`Inline` edit mode]({%slug treelist-editing-inline%}) and [`Popup` edit mode]({%slug treelist-editing-popup%}) integrate with cell selection without limitations.

### Selection and Virtual Scrolling

When the TreeList has [virtual scrolling]({%slug treelist-virtual-scrolling%}), the component is able to select a range of cells with **Shift** only if all rows in that range are currently rendered. Consider the following scenario:

1. Select a cell.
1. Scroll down, so that virtualization kicks in and the rendered rows are no longer the same.
1. Select another cell with **Shift**.

In this case, the range selection will start from the first row that is currently rendered. Compare with [Selection and paging](#selection-and-paging) below.

### Selection and Paging

The `SelectedCells` collection persists across paging.

### Selection and Templates

When using [TreeList templates]({%slug treelist-templates-overview%}) with cell selection:

> * If you are using a [TreeList column template]({%slug treelist-templates-column%}) and you have a clickable element in the template, wrap this element in a container with a `@onclick:stopPropagation` directive. You can check the knowledge base article on [how to prevent row selection when the user clicks another component in the Grid column template]({%slug grid-kb-row-selection-in-column-template%}). It is for the Grid component, but the concept is identical and it applies for both row and cell selection.

Reviewer: ```suggestion
* If you are using a [TreeList column template]({%slug treelist-templates-column%}) and you have a clickable element in the template, wrap this element in a container with an `@onclick:stopPropagation` directive. You can check the knowledge base article on [how to prevent row selection when the user clicks another component in the Grid column template]({%slug grid-kb-row-selection-in-column-template%}). It is for the Grid component, but the concept is identical and it applies to both row and cell selection.
```
